Replace Connection::query_all with Connection::query_iter

piodul opened the pull request:
For internal queries that can return a large number of rows (topology info, schema) we use currently use `Connection::query_all` to perform them. That method fetches results of the query in multiple pages and then returns it as a single combined `QueryResult`. This PR replaces it with a new `Connection::query_iter` method which has semantics more similar to the iterator-based query methods from the public interface (`Session::query_iter`, `Session::execute_iter`).

The motivations for this change are as follows:

- Although the new methods will use slightly more memory if the query returns only one page (because of the need to spawn a new task/channels/etc.), it should consume less memory if multiple pages are fetched because rows are processed on-line as new pages appear and only 1-2 pages need to exist at a given time.
- The logic that combines multiple pages into a single `QueryResult` will be hard to translate to use the upcoming iterator-based deserialization interface (https://github.com/scylladb/scylla-rust-driver/issues/462). Although the old interface will still be available as a deprecated fallback, I would like to adapt those methods as a form of dogfooding test for the new interface.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~I added appropriate `Fixes:` annotations to PR description.~

piodul commented:
```
running 2 tests
test retries_occur ... ok
test speculative_execution_is_fired ... FAILED

failures:

---- speculative_execution_is_fired stdout ----
Unique name: test_rust_1676631776_1
thread '<unnamed>' panicked at 'Worker failed: Node 172.42.0.2:9042 disconnected', scylla/tests/retries.rs:95:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'speculative_execution_is_fired' panicked at 'timeout: the function call took 3772 ms. Max time 30000 ms', scylla/tests/retries.rs:20:1


failures:
    speculative_execution_is_fired

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.77s
```

Strange, sounds a little bit as if a node has crashed. It's hard to tell for sure because we don't print logs from the scylla docker containers at all.

I'll try to improve the CI so that it does print the logs from the cluster. In the meantime, I'll just re-run and hope that the error was an unimportant fluke.

wprzytula commented on scylla/src/transport/connection.rs:
```diff
@@ -1559,8 +1558,7 @@ mod tests {
     ///    Then use query_all with page_size set to 7 to select all 100 rows.
-    /// 3. INSERT query_all should have None in result rows.
-    /// 4. Calling query_all with a Query that doesn't have page_size set should result in an error.
+    /// 3. INSERT query_all should work and not return any rows.
```
I can't see any commit replacing refs to `query_all` in comments with refs to `query_iter`.

piodul replied:
Done (in the commit which adjusts the test to use query_iter).

wprzytula commented on scylla/src/transport/connection.rs:
```diff
@@ -1559,8 +1558,7 @@ mod tests {
-    /// 4. Calling query_all with a Query that doesn't have page_size set should result in an error.
+    /// 3. INSERT query_all should work and not return any rows.
     #[tokio::test]
     async fn connection_query_all_test() {
```
Ditto with the test name: shouldn't it change its to name to `connection_query_iter_test` ? I understand that it still tests querying _all_ entities of some kind, but the function we test here is `query_iter`, so we should rather keep the name consistent with the function tested.

piodul replied:
Yeah, I just forgot to update the comments and the tests name. It doesn't make sense to have `query_all` anywhere in the repo at this point.

piodul replied:
Fixed, I also replaced the `query_all` substrings in the names of the tables created during the test.

wprzytula reviewed (changes requested):
Very good overall, but some old naming is still left.

piodul commented on scylla/src/transport/iterator.rs:
```diff
@@ -683,6 +715,66 @@ where
+/// A massively simplified version of the RowIteratorWorker. It does not have
+/// any complicated logic related to retries, it just fetches pages from
+/// a single connection.
+struct SingleConnectionRowIteratorWorker<Fetcher> {
+    sender: ProvingSender<Result<ReceivedPage, QueryError>>,
+    fetcher: Fetcher,
+}
+
+impl<Fetcher, FetchFut> SingleConnectionRowIteratorWorker<Fetcher>
+where
+    Fetcher: Fn(Option<Bytes>) -> FetchFut + Send + Sync,
```
Yes, it is conventional not to place constraints in type definitions unless they are absolutely necessary. I learned about this convention while learning Haskell, but when I usually look at the struct definitions in the Rust ecosystem I can see the same principle is used, too.

There are some good arguments in favor of this convention in this SO answer: https://stackoverflow.com/a/66369912

piodul commented on scylla/src/transport/topology.rs:
```diff
@@ -510,42 +510,40 @@ async fn query_peers(
 
-    let peers = try_join_all(translated_peers_futures).await?;
+    let peers = translated_peers_futures
+        .buffer_unordered(64)
```
It's an intuitive guess, there is no deep reasoning behind it... The `buffer_unordered` method required me to choose some number.

This number will limit the number of concurrent address translations (previously it could be unbounded). I don't think it is a bad idea to limit it, though it usually won't matter as `64` is higher than most cluster sizes so it won't effectively change the old behavior for most users. I can add a comment that this number is just a guess.

Do you have a suggestion for a better number?

wprzytula replied:
I would aim to choose a number that would be an upper size bound for a vast majority of clusters. If it were me who were to guess, it would be 256, but 64 is IMO OK too.

piodul replied:
All right, I changed to 256. I was a bit worried that it would increase memory usage, but the implementation doesn't seem to pre-allocate anything.

piodul commented:
v2:
- Changed the `buffer_unordered` parameter from `64` -> `256` to allow more concurrency when translating addresses,
- Updated some leftover comments/function names/string literals not to mention `query_all` but rather `query_iter`.

piodul commented:
@cvybhu ping

cvybhu commented on scylla/src/transport/iterator.rs:
```diff
@@ -683,6 +715,66 @@ where
     }
 }
 
+/// A massively simplified version of the RowIteratorWorker. It does not have
+/// any complicated logic related to retries, it just fetches pages from
+/// a single connection.
```
It would be nice to use a `RetryPolicy` for retries. I think it wouldn't be too hard to incorporate it.

piodul replied:
I'm not sure whether it makes sense for queries that are limited to a single connection and only read data from a single node. For example: what should happen on `RetryNextNode`?

cvybhu reviewed (commented):
Code looks ok, left a few nits.

Given that, I'm not a fan of this change. The streams are very fancy, but IMO they are much harder to read and reason about. Previously there were a few clearly defined steps where one thing happened at a time and I could easily follow it step after step. With streams it feels different - it's like first we prepare what to do and then do everything at once.

With normal code there are usually a few intermediate variables with clearly defined types, but here all the chained stream combinators make it impossible to be sure about the type of things at each step. The error handling is also much more complicated, all the errors must be propagated in the stream, and then all the `try_something` functions hopefully handle them, but I find it very hard to reason about. Then there are also the complications with `.buffer_unordered(256)`

IMO it would be better to keep using the `query_all` function. It could be implemented using `query_iter`, although I guess it doesn't make much sense.
I'm a big fan of KISS.

I might be complaining too much, maybe streams are just a Rust thing that I'm not used to yet, but my initial feeling is that they introduce new complexity that could easily be avoided.

In the PR description you mentioned that it will be hard to keep `query_all` after the deserialization refactor, but why is that? Will it be impossible to simply collect the rows to a Vec?

piodul commented:
> Code looks ok, left a few nits.
> 
> Given that, I'm not a fan of this change. The streams are very fancy, but IMO they are much harder to read and reason about. Previously there were a few clearly defined steps where one thing happened at a time and I could easily follow it step after step. With streams it feels different - it's like first we prepare what to do and then do everything at once.

I assume that your comment is about the changes in `query_peers` from `topology.rs`.

I'm not sure if I agree. The biggest advantage of streams here is that they reduce the peak memory use. Instead of loading all results from a query into memory, we only need to keep 1-2 pages in memory at a given time. I'll admit that the benefit is not huge because we are constructing the metadata anyway, but it's not only about making the code look more "fancy".

> 
> With normal code there are usually a few intermediate variables with clearly defined types, but here all the chained stream combinators make it impossible to be sure about the type of things at each step. The error handling is also much more complicated, all the errors must be propagated in the stream, and then all the `try_something` functions hopefully handle them, but I find it very hard to reason about. Then there are also the complications with `.buffer_unordered(256)`

We are already using iterators extensively in the codebase, so I thought that streams would be familiar. They have a slightly different API due to their asynchronicity, but the concept is the same.

I'm not sure what do you mean with complications with `.buffer_unordered()`. Is it about the need to explicitly limit parallelism? Limiting parallelism is not a bad thing.

> 
> IMO it would be better to keep using the `query_all` function. It could be implemented using `query_iter`, although I guess it doesn't make much sense. I'm a big fan of KISS.

With `query_iter`, you can choose either to collect everything to a Vec and then process, or do the processing on the fly, as pages are fetched.

> 
> I might be complaining too much, maybe streams are just a Rust thing that I'm not used to yet, but my initial feeling is that they introduce new complexity that could easily be avoided.
> 
> In the PR description you mentioned that it will be hard to keep `query_all` after the deserialization refactor, but why is that? Will it be impossible to simply collect the rows to a Vec?

I guess it will be possible to change query_all to just return a Vec of parsed rows, not a full QueryResponse. I just felt that solving the problem in this way will prevent us from utilizing the new interface to its full potential and will deny us from testing its ergonomics on our internal code.

In general, the new deserialization interface will not return pre-parsed rows in the style of `Vec<Vec<Option<CqlValue>>>>`, but will return an iterator that lazily deserializes the contents of a response. If we wanted the `query_all` to keep returning QueryResponse, then we would have to concatenate the contents of the pages into a large blob, which is even worse with regards to memory use.

piodul commented on scylla/src/transport/iterator.rs:
```diff
@@ -265,6 +266,37 @@ impl RowIterator {
         Self::new_from_worker_future(worker_task, receiver).await
     }
 
+    pub(crate) async fn new_for_query_iter(
```
Good point, changed to `new_for_connection_query_iter`.

piodul commented on scylla/src/transport/iterator.rs:
```diff
@@ -265,6 +266,37 @@ impl RowIterator {
+                fetcher: |paging_state| {
+                    connection.query_with_consistency(
+                        &query,
+                        values.clone(),
```
You are right, it wasn't needed. Changed.

piodul commented:
v3:
- renamed `new_for_query_iter` -> `new_for_connection_query_iter`
- got rid of an unnecessary `.clone()` in a lambda defined in `RowIterator::new_for_connection_query_iter`

piodul commented:
@cvybhu review ping
